src/cfg/domtree.h

```diff
@@ -0,0 +1,178 @@
+  // up this value at the very end). All other nodes start with a nonsense value
+  // that indicates they have yet to be processed.
+  iDoms.resize(numBlocks, nonsense);
+  iDoms[0] = 0;
+
+  // Process the (non-entry) blocks in reverse postorder, computing the
+  // immediate dominators as we go. This returns whether we made any changes,
+  // which is used in an assertion later down.
+  auto processBlocks = [&]() {
+    bool changed = false;
+    for (Index index = 1; index < numBlocks; index++) {
+      // Loop over the predecessors. Our new parent is basically the
+      // intersection of all of theirs: our immediate dominator must precede all
+      // of them.
+      auto& preds = blocks[index]->in;
+      Index newParent = nonsense;
+      for (auto* pred : preds) {
+        auto predIndex = blockIndices[pred];
+
+        // In a reducible graph, we only need to care about the predecessors
+        // that appear before us in the reverse postorder numbering. The only
+        // predecessor that can appear *after* us is a loop backedge, but that
+        // will never dominate the loop - the loop is dominated by its single
+        // entry (since it is reducible, it has just one entry).
+        if (predIndex > index) {
+          continue;
+        }
+
+        // All of our predecessors will have been processed before us, except
+        // if they are unreachable from the entry, in which case, we can ignore
+        // them.
+        if (iDoms[predIndex] == nonsense) {
+          continue;
+        }
+
+        if (newParent == nonsense) {
+          // This is the first processed predecessor.
+          newParent = predIndex;
+          continue;
+        }
+
+        // This is an additional predecessor. Intersect it, by going back to a
+        // node that definitely dominates both possibilities. Effectively, we
+        // keep decreasing the index backwards in the reverse postorder
+        // indexing until we stop (at the latest, in the entry).
+        auto left = newParent;
+        auto right = predIndex;
+        while (left != right) {
+          while (left > right) {
+            left = iDoms[left];
+          }
+          while (right > left) {
+            right = iDoms[right];
+          }
+        }
+        newParent = left;
+      }
+
+      // Check if we found a new value here, and apply it. (We will normally
+      // always find a new value in the single pass that we run, but we also
+      // assert lower down that running another pass causes no further changes.)
+      if (newParent != iDoms[index]) {
+        iDoms[index] = newParent;
+        changed = true;
+
+        // In reverse postorder the dominator cannot appear later.
+        assert(newParent <= index);
+      }
+    }
+
+    return changed;
+  };
+
+  processBlocks();
+
+  // We must have finished all the work in a single traversal, since our input
+  // is reducible.
+  assert(!processBlocks());
```

Comment:
Do we ship release builds with assertions or without? If we ship with them, isn't this gonna run the process twice unnecessarily?

Reply:
By default we build with assertions, but I think @dschuff was looking into disabling that in LTO builds in the emsdk, or maybe we already do?

Reply:
Maybe we can do
```cpp
BYN_DEBUG(!processBlocks());
```
using [BYN_DEBUG](https://github.com/WebAssembly/binaryen/blob/a2323f2cfd90089c54100ab98c439b9438cc4dc1/src/support/debug.h#L58)? It only runs with debug mode. Disabling assertions altogether may have some consequences, but running the whole process once again may be too high a cost here..

Reply:
Interesting idea. However, I think we rarely build with `BYN_DEBUG` though - we don't test it by default anywhere, do we? And the overhead here is quite small, as this will run in 2 passes instead of 1, but the original algorithm would take potentially many more passes, but is still considered one of the fastest in this area. That feels like a reasonable amount of overhead to me in a build with assertions.